File: oifs_43r3_test.py
#! /usr/bin/python3

# Script to simulate an oifs_43r3 executable for testing purposes

# This script has been written by Andy Bowery (Oxford University, 2023)

if __name__ == "__main__":

    import argparse, time, secrets
    from datetime import datetime

    # This script is run from the command line using: ./oifs_43r3_test.py --nfrres=N --upload_interval=N


    # Parse the command line, use argparse to read in the options from the shell command line
    parser = argparse.ArgumentParser()
    parser.add_argument("--nfrres",default=10000)
    parser.add_argument("--upload_interval",default=10000)
    options = parser.parse_args()


    print("Starting oifs_43r3_test: "+str(datetime.now()))

    time.sleep(2)

    iteration = 0
    max_iter = 5

    #options.upload_interval = 2


    while iteration <= max_iter:

       if iteration % abs( int(options.nfrres) ) == 0:
          iteration2 = 3
       else:
          iteration2 = 1

       # Write to the ifs.stat file
       for x in range(iteration2):
          with open('ifs.stat', 'a') as ifs_stat_file:
             if len(str(iteration)) == 1:
                ifs_stat_file.write(' ' + datetime.now().strftime("%H:%M:%S") + ' 0AAA00AAA STEPO       ' + str(iteration) + '\n')
             if len(str(iteration)) == 2:
                ifs_stat_file.write(' ' + datetime.now().strftime("%H:%M:%S") + ' 0AAA00AAA STEPO      ' + str(iteration) + '\n')


       # Write out the ICM files if at the end of an upload_interval
       if iteration % int(options.upload_interval) == 0 and iteration > 0:
          if len(str(iteration)) == 1:
             second_part = '00000' + str(iteration)
          elif len(str(iteration)) == 2:
             second_part = '0000' + str(iteration)
          elif len(str(iteration)) == 3:
             second_part = '000' + str(iteration)

          # Write to the ICMGG
          with open('ICMGG' + second_part, 'a') as ICMGG_file:
             ICMGG_file.write(secrets.token_hex(4000) + '\n')

          # Write to the ICMSH file
          with open('ICMSH' + second_part, 'a') as ICMSH_file:
             ICMSH_file.write(secrets.token_hex(4000) + '\n')

          # Write to the ICMUA file
          with open('ICMUA' + second_part, 'a') as ICMUA_file:
             ICMUA_file.write(secrets.token_hex(4000) + '\n')

       iteration = iteration + 1



    # And finally write the last CNTO line into the ifs.stat file
    with open('ifs.stat', 'a') as ifs_stat_file:
       if len(str(iteration)) == 1:
          ifs_stat_file.write(' ' + datetime.now().strftime("%H:%M:%S") + ' 0AAA00AAA CNTO        ' + str(iteration-1) + '\n')
       if len(str(iteration)) == 2:
          ifs_stat_file.write(' ' + datetime.now().strftime("%H:%M:%S") + ' 0AAA00AAA CNTO       ' + str(iteration-1) + '\n')


    # Produce NODE file
    with open('NODE.001_01', 'a') as node_file:
       node_file.write(secrets.token_hex(4000) + '\n')

File: oifs_43r3_test.cpp
// oifs_43r3_test.exe

// Program to simulate an oifs_43r3 executable for testing purposes
// This program has been written by Andy Bowery (Oxford University, 2023)

// To build use: g++ oifs_43r3_test.cpp -std=c++17 -o oifs_43r3_test.exe

// This program is run from the command line using: ./oifs_43r3_test.exe

#include <iostream>
#include <iomanip>
#include <ctime>
#include <sstream>
#include <stdlib.h>
#include <stdio.h>
#include <fstream>
#include <filesystem>
#include <string>
#include <chrono>
#include <thread>

using namespace std;
using namespace std::chrono;


int main()
{
    std::string second_part, exptid = "EXPT";
    int iteration = 0;
    int iteration2 = 0;
    int max_iter = 24;     // max number of timesteps  !! TODO: These need to come from a mock namelist !!!

    int nfrres = 2; // restart interval
    int upload_interval = 4;

    cerr << "Starting oifs_43r3_test" << std::endl;

    // Get the slots path (the current working path)
    std::string slot_path = std::filesystem::current_path();

    std::string ifs_stat_file = slot_path + std::string("/ifs.stat");
    std::ofstream ifs_stat_file_out(ifs_stat_file);

    auto t = std::time(nullptr);
    auto tm = *std::localtime(&t);

    this_thread::sleep_until(system_clock::now() + seconds(10));

    while (iteration <= max_iter) {

       // At the end of every restart interval, write out the same line three times
       if ( iteration % abs( nfrres ) == 0) {
          iteration2 = 3;
       } else {
          iteration2 = 1;
       }

       // Write to the ifs.stat file
       for (auto i=0; i < iteration2; i++) {
          if ( to_string(iteration).length() == 1) {
             ifs_stat_file_out <<" "<< std::put_time(&tm, "%H:%M:%S") << " 0AAA00AAA STEPO       " << to_string(iteration) << std::endl;
             cerr              <<" "<< std::put_time(&tm, "%H:%M:%S") << " 0AAA00AAA STEPO       " << to_string(iteration) << std::endl;
          } else if ( to_string(iteration).length() == 2) {
             ifs_stat_file_out <<" "<< std::put_time(&tm, "%H:%M:%S") << " 0AAA00AAA STEPO      " << to_string(iteration) << std::endl;
             cerr              <<" "<< std::put_time(&tm, "%H:%M:%S") << " 0AAA00AAA STEPO      " << to_string(iteration) << std::endl;
          }
       }

       // Write out the ICM files if at the end of an upload_interval
       if (iteration % upload_interval == 0 and iteration > 0) {
          if ( to_string(iteration).length() == 1) {
             second_part = exptid +"+"+ "00000" + to_string(iteration);
          } else if ( to_string(iteration).length() == 2) {
             second_part = exptid +"+"+ "0000" + to_string(iteration);
          } else if ( to_string(iteration).length() == 3) {
             second_part = exptid +"+"+ "000" + to_string(iteration);
          }

          std::string ICMGG_file = slot_path + std::string("/ICMGG") + second_part;
          std::string ICMSH_file = slot_path + std::string("/ICMSH") + second_part;
          std::string ICMUA_file = slot_path + std::string("/ICMUA") + second_part;

          // Open the ICM file streams
          std::ofstream ICMGG_file_out(ICMGG_file);
          std::ofstream ICMSH_file_out(ICMSH_file);
          std::ofstream ICMUA_file_out(ICMUA_file);

          // Write to the ICMGG
          for (auto j=0; j<4000; j++) { ICMGG_file_out << rand() % 10; };
          ICMGG_file_out << std::endl;

          // Write to the ICMSH
          for (auto j=0; j<4000; j++) { ICMSH_file_out << rand() % 10; };
          ICMSH_file_out << std::endl;

          // Write to the ICMUA
          for (auto j=0; j<4000; j++) { ICMUA_file_out << rand() % 10; };
          ICMUA_file_out << std::endl;

          // Close the ICM file streams
          ICMGG_file_out.close();
          ICMSH_file_out.close();
          ICMUA_file_out.close();

       }

       iteration = iteration + 1;

       // Time delay to slow the program down to allow the main loop of the calling program to run
       this_thread::sleep_until(system_clock::now() + seconds(10));
    }


    // And finally write the last CNT0 line into the ifs.stat file
    if ( to_string(iteration).length() == 1) {
       ifs_stat_file_out <<" "<< std::put_time(&tm, "%H:%M:%S") << " 0AAA00AAA CNT0        " << to_string(iteration) << std::endl;
       cerr              <<" "<< std::put_time(&tm, "%H:%M:%S") << " 0AAA00AAA CNT0        " << to_string(iteration) << std::endl;
    } else if ( to_string(iteration).length() == 2) {
       ifs_stat_file_out <<" "<< std::put_time(&tm, "%H:%M:%S") << " 0AAA00AAA CNT0       " << to_string(iteration) << std::endl;
       cerr              <<" "<< std::put_time(&tm, "%H:%M:%S") << " 0AAA00AAA CNT0       " << to_string(iteration) << std::endl;
    }
    ifs_stat_file_out.close();


    // Produce the NODE file
    std::string NODE_file = slot_path + std::string("/NODE.001_01");
    std::ofstream NODE_file_out(NODE_file);
    for (auto j=0; j<4000; j++) {
      NODE_file_out << rand() % 10 << '\n';
    }
    NODE_file_out << std::endl;
    NODE_file_out.close();

}
